File: leetcode/character-replacement.go
package leetcode

func characterReplacement(s string, k int) int {
	countMap := make(map[byte]int)
	l, res, maxFreq := 0, 0, 0
	for r := 0; r < len(s); r++ {
		countMap[s[r]]++
		if countMap[s[r]] > maxFreq {
			maxFreq = countMap[s[r]]
		}

		if (r-l+1)-maxFreq > k {
			countMap[s[l]]--
			l++
		}

		if (r - l + 1) > res {
			res = r - l + 1
		}
	}
	return res
}

File: leetcode/character-replacement.ts
function characterReplacement(s: string, k: number): number {
  const countMap: Record<string, number> = {};
  let l = 0,
    maxFreq = 0,
    res = 0;

  for (let i = 0; i < s.length; i++) {
    countMap[s[i]] = (countMap[s[i]] || 0) + 1;

    if (countMap[s[i]] > maxFreq) {
      maxFreq = countMap[s[i]];
    }

    if (i - l + 1 - maxFreq > k) {
      countMap[s[l]]--;
      l++;
    }

    if (i - l + 1 > res) {
      res = i - l + 1;
    }
  }

  return res;
}
